### ISSUE TITLE
i18n: Configure minification to retain translator comments

### ISSUE THREAD

**Issue by aduth** [WordPress/gutenberg#19006]
Currently, translator comments are not represented for the plugin for JavaScript code, due to the fact that the plugin is shipped with minified code in which the inline code comments have been removed.

https://github.com/WordPress/gutenberg/blob/8f649f23d6570bbfc46eae54a5ca799edc744396/packages/block-editor/src/components/multi-selection-inspector/index.js#L29-L30

You can see that the plugin strings filtered to those with comments include only localized strings found in the PHP source:

https://translate.wordpress.org/projects/wp-plugins/gutenberg/dev/en-ca/default/?filters%5Bterm%5D=&filters%5Buser_login%5D=&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filters%5Bwith_comment%5D=yes&filter=Filter&sort%5Bby%5D=priority&sort%5Bhow%5D=desc

**Task:** Avoid removing translator comments during JavaScript minification, or otherwise make available the original source files that can be used for string extraction in the plugin repository.

**Implementation Specifics:**

Webpack uses Terser for its default minification ([reference](https://webpack.js.org/configuration/optimization/)). Terser provides a `comments` option which can be provided with a regular expression of comments to preserve ([reference](https://github.com/terser/terser#output-options)). We could provide a regular expression which matches the translator comment syntax to avoid removing the comments. This can be configured by specifying a custom `optimization.minifier` instance of `TerserPlugin` ([reference](https://webpack.js.org/configuration/optimization/#optimizationminimizer)). Given that this would be a typical expectation for any plugin leveraging the default Webpack configuration, I would propose it to be included in the shared configuration provided through `@wordpress/scripts` ([source](https://github.com/WordPress/gutenberg/blob/master/packages/scripts/config/webpack.config.js)).

**Open Questions:**

A potential downside of including these comments in the minified output is that, aside from the static analysis performed for string extraction, these comments do not provide any meaningful value to the application runtime and thus can be considered wasted bytes.

Alternatively, if it would be enough to provide the original source files as part of the plugin distributable, this may be a viable option. However, it's unclear if the "duplication" of the strings between the minified output and the raw source can be reliably reconciled by the plugin repository's string extraction.

cc @swissspidy

**Comment by swissspidy** [comment]
> A potential downside of including these comments in the minified output is that, aside from the static analysis performed for string extraction, these comments do not provide any meaningful value to the application runtime and thus can be considered wasted bytes.

I don't really see this as a huge issue as the overhead of some translator comments is negligible.

> Alternatively, if it would be enough to provide the original source files as part of the plugin distributable, this may be a viable option. However, it's unclear if the "duplication" of the strings between the minified output and the raw source can be reliably reconciled by the plugin repository's string extraction.

Duplication is not an issue. The system handles strings from multiple sources just fine.

For translators, having the actual source files would theoretically be a huge improvement. But that would mean shipping tons of unneeded code to plugin users. _This_ I would consider wasted bytes :-) It might also cause issues on dotorg infra.

**Comment by aduth** [comment]
> > A potential downside of including these comments in the minified output is that, aside from the static analysis performed for string extraction, these comments do not provide any meaningful value to the application runtime and thus can be considered wasted bytes.
> 
> I don't really see this as a huge issue as the overhead of some translator comments is negligible.

Sure, but when it comes to performance impact of accruing bundle sizes, it's almost always of the ["death by a thousand cuts"](https://en.wikipedia.org/wiki/Death_by_a_thousand_cuts_(psychology)) sort. Weighed to the alternative, I'd grant it could still be the best compromise, but for exhaustiveness' sake, I felt it was worth note.

>Duplication is not an issue. The system handles strings from multiple sources just fine.

For me, it's not clear which would be treated as "the truth", if one occurrence of the string has an associated translator comment, and the other does not. Will it always merge to use the translator comment?

>For translators, having the actual source files would theoretically be a huge improvement. But that would mean shipping tons of unneeded code to plugin users. This I would consider wasted bytes :-) It might also cause issues on dotorg infra.

I have no sense of what precisely this amounts to, and whether it would actually be a problem. I'm also curious what we need to be doing here to assure that the translator comments are present for the block editor strings _in core_, as well as in the plugin. I imagine we would want to avoid any situation where the source files would be shipped with core, but I'm also not sure if that string extraction operates on the built copy of WordPress, on `wordpress-develop`, or perhaps even inheriting from the Gutenberg plugin in some way? I'm not too familiar with this process.

**Comment by aduth** [comment]
Some recent discussion on this issue in Slack ([links require registration](https://make.wordpress.org/chat/)):

https://wordpress.slack.com/archives/C02QB2JS7/p1578592522003300
https://wordpress.slack.com/archives/C02QB2JS7/p1579100912175100

The most recent conversation touches a bit more on: Even if we include these translator comments in the minified output, it's still not ideal that the code references that are presented to a translator would direct them to minified code. It's assumed that these code references could be useful for a translator to provide context for how the strings are used in the application code.

@swissspidy touched on this as well at https://github.com/WordPress/gutenberg/issues/19006#issuecomment-564334273 :

>For translators, having the actual source files would theoretically be a huge improvement.

I'm not sure what an ideal solution would be here. Is there an opportunity to explore something akin to a [source map](https://developer.mozilla.org/en-US/docs/Tools/Debugger/How_to/Use_a_source_map) or some other link to a "source of truth". And could that source of truth might be remote (i.e. a link to the GitHub source repository, rather than requiring some relative file within the same plugin directory)?

**Comment by swissspidy** [comment]
>  Even if we include these translator comments in the minified output, it's still not ideal that the code references that are presented to a translator would direct them to minified code. It's assumed that these code references could be useful for a translator to provide context for how the strings are used in the application code.

That's a separate issue, really.

I was just translating the string `(%s: color %s)` from the block editor package and had no clue what the placeholder meant. Keeping translator comments would already help me quite a bit. Sure, the source code reference could be better, but not a requirement in this scenario.

> I'm not sure what an ideal solution would be here. Is there an opportunity to explore something akin to a source map or some other link to a "source of truth". 

Perhaps? But that would probably be quite a big undertaking and not something we could implement over night. Perhaps @ocean90 can chime in here on the possibilites for that on dotorg.

**tl;dr**: Adding translator comments would already help a ton. Actually linking to the source code would be even better, but is a separate enhancement.
